warehouse/query-core/src/main/java/datawave/query/iterator/logic/AndIterator.java

```diff
@@ -73,220 +107,239 @@ public AndIterator(Iterable<NestedIterator<T>> sources, Iterable<NestedIterator<
+            } else if (includeHeads.keySet().contains(highest)) {
+                // need to advance all sources that map to the highest key
+                SortedSet<NestedIterator<T>> sources = includeHeads.removeAll(highest);
+                for (NestedIterator<T> source : sources) {
+                    if (source.hasNext()) {
+                        T result = source.next();
+                        if (result != null) {
+                            T transformed = transformer.transform(result);
+                            includeHeads.put(transformed, source);
+                            // optimization, if the returned element is beyond the highest element, reset highest
+                            if (transformed.compareTo(highest) > 0) {
+                                highest = transformed;
+                            }
+                        } else if (!Util.isIteratorExhausted(source)) {
+                            // an include source can expire in a situation like
+                            // CONTEXT && (A || (B && !C)) which will end up having an expired iterator
+                            expiredIncludes.add(source);
+                        } else {
+                            // iterator exhausted, we're done
+                            exhaustedIncludes.addAll(includeHeads.values());
+                            exhaustedIncludes.addAll(sources);
+                            exhaustedIncludes.addAll(expiredIncludes);
+                            includeHeads.clear();
+                            return;
+                        }
+                    } else {
+                        // iterator exhausted, we're done
+                        exhaustedIncludes.addAll(includeHeads.values());
+                        exhaustedIncludes.addAll(sources);
+                        exhaustedIncludes.addAll(expiredIncludes);
+                        includeHeads.clear();
+                        return;
+                    }
                 }
             }
-        }
 
-        // if we didn't move after the loop, then we don't have a next after this
-        if (prev == next) {
-            next = null;
+            // moving expired sources must happen last. Otherwise, we could move into a valid hit
+            // and immediately call next on the sources.
+            moveExpiredIncludeSources(highest);
         }
-
-        return prev;
     }
 
+    @Override
     public void remove() {
         throw new UnsupportedOperationException("This iterator does not support remove.");
     }
 
-    public boolean hasNext() {
-        if (!isInitialized()) {
-            throw new IllegalStateException("initialize() was never called");
+    /**
+     * Handles removing an iterator from the underlying collections if an exception occurs.
+     *
+     * @param iterator
+     *            the iterator to remove
+     */
+    private void removeIterator(NestedIterator<T> iterator) {
+        log.error("removing iterator");
```

Comment:
Should this be a different log level? And maybe include some info to say which iterator is being removed to aid in debugging?